matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -297,33 +420,75 @@ private void computeTransitionProbabilities(TimeStep<GPXExtension, GPXEntry, Pat
-                final Path path = algo.calcPath(from.getQueryResult().getClosestNode(),
-                        to.getQueryResult().getClosestNode());
+
+                final Path path = algo.calcPath(from.queryResult.getClosestNode(),
+                        to.queryResult.getClosestNode());
+
                 if (path.isFound()) {
                     timeStep.addRoadPath(from, to, path);
+
+                    // The router considers unfavored virtual edges using edge penalties.
+                    // However, if a virtual edge is included in the computed path,
+                    // the path distance does not include any penalty distance. Hence, we
+                    // need to add our own penalty distance for every unfavored edge in the path.
+                    final double penalizedPathDistance = penalizedPathDistance(path,
+                            queryGraph.getUnfavoredVirtualEdges());
+
+                    logger.info("Path from: {}, to: {}, penalized path length: {}",
+                            from, to, penalizedPathDistance);
+
                     final double transitionLogProbability = probabilities
-                            .transitionLogProbability(path.getDistance(), linearDistance, timeDiff);
+                            .transitionLogProbability(penalizedPathDistance, linearDistance);
                     timeStep.addTransitionLogProbability(from, to, transitionLogProbability);
+                } else {
+                    logger.info("No path found for from: {}, to: {}", from, to);
                 }
+                queryGraph.clearUnfavoredStatus();
+
             }
         }
     }
 
-    private MatchResult computeMatchResult(List<SequenceState<GPXExtension, GPXEntry, Path>> seq,
-                                           List<GPXEntry> gpxList, List<QueryResult> allCandidates,
-                                           EdgeExplorer explorer) {
-        // every virtual edge maps to its real edge where the orientation is already correct!
-        // TODO use traversal key instead of string!
-        final Map<String, EdgeIteratorState> virtualEdgesMap = new HashMap<>();
-        for (QueryResult candidate : allCandidates) {
-            fillVirtualEdges(virtualEdgesMap, explorer, candidate);
+    /**
+     * Returns the path length plus a penalty for each used unfavored edge.
+     */
+    private double penalizedPathDistance(Path path,
+                                         List<VirtualEdgeIteratorState> penalizedVirtualEdges) {
+        final Set<Integer> penalizedEdges = new HashSet<>();
+        for (VirtualEdgeIteratorState edge : penalizedVirtualEdges) {
+            penalizedEdges.add(edge.getEdge());
         }
+        double totalPenalty = 0;
+        for (EdgeIteratorState edge : path.calcEdges()) {
+            if (penalizedEdges.contains(edge.getEdge())) {
+                totalPenalty += VIRTUAL_NODE_U_TURN_PENALTY;
+            }
+        }
```

Comment:
My point was more that we'll only ever have 2 (or 4?) entries in the set, and only calling 'contains' twice (i.e. first/last edge of path), so I'm not sure we need a hashset. (And, if we are using one, we might as well just hash the edge IDs, hence TIntHashSet. Or we could even use int[].) However, if it's premature optimization, I'm more than happy to ignore it.